plugins/askrene/refine.c

```diff
@@ -0,0 +1,379 @@
+			   decreased == -1 ? "none"
+			   : fmt_flows_step_scid(tmpctx, rq, flow, decreased));
+		flow->delivers = msat;
+	}
+
+	return NULL;
+}
+
+/* Flow is now delivering `extra` additional msat, so modify reservations */
+static void add_to_flow(struct flow *flow,
+			struct route_query *rq,
+			struct reservations *reservations,
+			struct amount_msat extra)
+{
+	struct amount_msat orig, updated;
+
+	orig = flow->delivers;
+	if (!amount_msat_add(&updated, orig, extra))
+		abort();
+
+	flow->delivers = updated;
+
+	/* Now add reservations accordingly (effects constraints on other flows)  */
+	for (int i = tal_count(flow->path) - 1; i >= 0; i--) {
+		const struct half_chan *h = flow_edge(flow, i);
+		struct amount_msat diff;
+
+		/* Can't happen, since updated >= orig */
+		if (!amount_msat_sub(&diff, updated, orig))
+			abort();
+		add_reservation(reservations, rq, flow, i, diff);
+
+		if (!amount_msat_add_fee(&orig, h->base_fee, h->proportional_fee))
+			abort();
+		if (!amount_msat_add_fee(&updated, h->base_fee, h->proportional_fee))
+			abort();
+	}
+}
+
+/* Check out remaining capacity for this flow.  Changes as other flows get
+ * increased (which sets reservations) */
+static struct amount_msat flow_remaining_capacity(struct route_query *rq,
+						  const struct flow *flow)
+{
+	struct amount_msat max_msat = AMOUNT_MSAT(-1ULL);
+	for (int i = tal_count(flow->path) - 1; i >= 0; i--) {
+		const struct half_chan *h = flow_edge(flow, i);
+		struct amount_msat min, max;
+
+		/* We can pass constraints due to addition of fees! */
+		get_constraints(rq, flow->path[i], flow->dirs[i], &min, &max);
+		max = amount_msat_min(max, amount_msat(fp16_to_u64(h->htlc_max)));
+
+		max_msat = amount_msat_min(max_msat, max);
+		if (!amount_msat_add_fee(&max_msat, h->base_fee, h->proportional_fee))
+			max_msat = AMOUNT_MSAT(-1ULL);
+	}
+
+	/* Calculate deliverable max */
+	for (size_t i = 0; i < tal_count(flow->path); i++) {
+		const struct half_chan *h = flow_edge(flow, i);
+		max_msat = amount_msat_sub_fee(max_msat,
+					       h->base_fee, h->proportional_fee);
+	}
+	return max_msat;
+}
+
+/* What's the "best" flow to add to? */
+static struct flow *pick_most_likely_flow(struct route_query *rq,
+					  struct flow **flows,
+					  struct amount_msat additional)
+{
+	double best_prob = 0;
+	struct flow *best_flow = NULL;
+
+	for (size_t i = 0; i < tal_count(flows); i++) {
+		struct amount_msat cap;
+		double prob = flow_probability(flows[i], rq);
+		if (prob < best_prob)
+			continue;
+		cap = flow_remaining_capacity(rq, flows[i]);
+		if (amount_msat_less(cap, additional))
+			continue;
+		best_prob = prob;
+		best_flow = flows[i];
+		plugin_log(rq->plugin, LOG_DBG, "Best flow is #%zu!", i);
+	}
+
+	return best_flow;
+}
+
+const char *
+refine_with_fees_and_limits(const tal_t *ctx,
+			    struct route_query *rq,
+			    struct amount_msat deliver,
+			    struct flow ***flows)
+{
+	struct reservations *reservations = new_reservations(NULL, rq);
```

Comment:
Why not allocate with `tmpctx`?

Reply:
We *have* to free it in this function (since we want them cleaned up).  This way, leak detection will notice if we don't, with tmpctx we would not know.